presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java

```diff
@@ -210,47 +212,47 @@ private Object getLegacyValue(Object value, Type type)
         }
 
         if (type instanceof ArrayType) {
+            checkArgument(value instanceof List, "ArrayType.getValueObject should return List");
+
             Type elementType = ((ArrayType) type).getElementType();
 
             if (!(elementType instanceof TimestampType || elementType instanceof TimestampWithTimeZoneType)) {
                 return value;
             }
 
-            return ((List<Object>) value).stream()
+            return unmodifiableList(((List<Object>) value).stream()
                     .map(element -> getLegacyValue(element, elementType))
-                    .collect(toImmutableList());
+                    .collect(toList()));
         }
 
         if (type instanceof MapType) {
+            checkArgument(value instanceof Map, "MapType.getValueObject should return Map");
+
             Type keyType = ((MapType) type).getKeyType();
             Type valueType = ((MapType) type).getValueType();
 
-            return ((Map<Object, Object>) value).entrySet().stream()
-                    .collect(toImmutableMap(entry -> getLegacyValue(entry.getKey(), keyType), entry -> getLegacyValue(entry.getValue(), valueType)));
+            // We cannot use streams here due to https://bugs.openjdk.java.net/browse/JDK-8148463
+            Map<Object, Object> result = new HashMap<>();
+            for (Map.Entry<Object, Object> entry : ((Map<Object, Object>) value).entrySet()) {
+                result.put(getLegacyValue(entry.getKey(), keyType), getLegacyValue(entry.getValue(), valueType));
+            }
+            return unmodifiableMap(result);
         }
 
         if (type instanceof RowType) {
-            List<RowType.Field> fields = ((RowType) type).getFields();
-
-            if (value instanceof Map) {
```

Comment:
Any idea why it was here before? Just an ommision?

Reply:
I guess so, I've double-checked and `RowType.getObjectValue` returns `List<Object>`. I've added assertions that will throw an exception if this is not true so we will know :)